node/lib/util/stopwatch.js

```diff
@@ -0,0 +1,163 @@
+/*
+ *
+ * Copyright (c) 2016, Two Sigma Open Source
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of git-meta nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+"use strict";
+
+const assert = require("chai").assert;
+
+/**
+ * This module provides the `Stopwatch` class.
+ */
+
+/**
+ * `Stopwatch` is a mechanism class providing a way to measure elapsed wall
+ * time.  `Stopwatch` supports *reentrant* code (as is common with 
+ * synchronous code), keeping so that after N calls to `start` it will
```

Comment:
I think I pulled it out of the places I was using it during profiling. E.g.:

```JavaScript
const totalTime = new Stopwatch(true);
yield someArray.map(co.wrap(function *(value) {
    // do some stuff
    //...
    // Now we're going to do the stuff I'm interested in timing.
    totalTime.start();
    // Do interesting stuff that may block.
    totalTime.stop();
    // Do more uninteresting stuff
});

console.log(`Spent ${totalTime.elapsed} doing interesting stuff.`);
```
We're going to end potentially calling `start` many times before matching calls to `stop` are seen. Without this counting capability the behavior of the second call to `start` is undefined (or if it does nothing, the very first call to `stop` will stop the timer). This pattern is pretty common in asynchronous code where there is a high amount of reentrancy within a single thread.